File: 20190104/17subsets.py
class Solution:
    """
    @param nums: A set of numbers
    @return: A list of lists
    """
    def subsets(self, nums):
        # write your code here

        # method 2
        result = []
        if nums is None:
            return result

        nums.sort()
        self.dfs(nums, 0, [], result)

        return result

    def dfs(self, nums, startIndex, subset, result):
        result.append(list(subset))

        for i in range(startIndex, len(nums)):
            subset.append(nums[i])
            self.dfs(nums, i + 1, subset, result)
            subset.pop()

    #     # method 1
    #     self.result = []
    #     self.search(sorted(nums), [], 0)

    #     return self.result

    # def search(self, nums, subset, index):
    #     if index == len(nums):
    #         self.result.append(list(subset))
    #         return

    #     subset.append(nums[index])
    #     self.search(nums, subset, index + 1)
    #     # Backtracking
    #     subset.pop()
    #     self.search(nums, subset, index + 1)

File: 06 - Implicit Graph DFS/17. Subsets.cpp
class Solution {
public:
    /**
     * @param nums: A set of numbers
     * @return: A list of lists
     */
    vector<vector<int>> subsets(vector<int> &nums) {
        // write your code here
        sort(nums.begin(), nums.end());

        // method 2
        // vector<vector<int>> result;
        // vector<int> subset;

        // dfs(nums, 0, subset, result);

        // return result;

        // method 1
        vector<vector<int>> result;
        vector<int> empty;
        result.push_back(empty);

        for (int num: nums) {
            int size = result.size();
            for (int i = 0; i < size; i++) {
                vector<int> subset = result[i];
                subset.push_back(num);
                result.push_back(subset);
            }
        }

        return result;
    }

    void dfs(vector<int> &nums, int index, vector<int> &subset, vector<vector<int>> &result) {
        result.push_back(subset);

        for (int i = index; i < nums.size(); i++) {
            subset.push_back(nums[i]);
            dfs(nums, i + 1, subset, result);
            subset.pop_back();
        }
    }
};
